### DISCUSSION TITLE
Change default tool

### DISCUSSION THREAD

**Discussion by salah-walid** [iTwin/itwinjs-core#3371] [Q&A]
Hello,

I am trying to change the default tool to the rotation tool, tried with the following code but it didn't work:
`IModelApp.toolAdmin.defaultToolId = "View.Rotate"; `
`await IModelApp.toolAdmin.startDefaultTool();`

**Comment by bbastings** [comment]
The default tool needs to be a subclass of PrimitiveTool as suggested by the method documentation:

  /** The name of the [[PrimitiveTool]] to use as the default tool. Defaults to "Select", referring to [[SelectionTool]].
   * @see [[startDefaultTool]] to activate the default tool.
   * @see [[defaultToolArgs]] to supply arguments when starting the tool.
   */
  public get defaultToolId(): string {
    return this._defaultToolId;
  }
  public set defaultToolId(toolId: string) {
    this._defaultToolId = toolId;
  }

This is due to the way a ViewTool only suspends the active PrimitiveTool while it is active.

You can however specify "" to have no default tool. In this case left button drag will be handled as view rotate by the default IdleTool.

HTH

**Reply by salah-walid** [reply]
I have set defaultToolId to an empty string and still left click doesn't rotate the view.
To be noted i am still using ITwin V2

**Comment by bbastings** [comment]
Left click or left drag? Left drag is mapped to rotate for the default IdleTool, click does nothing.

**Reply by salah-walid** [reply]
I meant left drag, it still uses selection drag

**Comment by bbastings** [comment]
Sounds like you didn't override the default tool early enough or it's being overridden by your frontstage?

**Reply by salah-walid** [reply]
I am overriding it in the viewportConfigurer event

**Reply by NancyMcCallB** [reply]
For now, in an itwin-viewer based application you need to use a hack like the following to override the default toolId baked into the viewer's default frontstage. I believe this is a short term issue and is likely being addressed in an upcoming release of the viewer. @aruniverse .



```
const onIModelAppInit = useCallback(() => {
MeasureLengthTool.register();
IModelApp.toolAdmin.defaultToolId = MeasureLengthTool.toolId
},[]);



useEffect(()=> {
const processFrontstageActivation = (args: FrontstageActivatedEventArgs) =>{
// clear out _defaultTool property so default tool is not set
Object.defineProperty(args.activatedFrontstageDef, "_defaultTool", {value: undefined});
IModelApp?.toolAdmin?.startDefaultTool();
}



<Viewer
iTwinId={iTwinId}
iModelId={iModelId}
authClient={authClient}
viewCreatorOptions={viewCreatorOptions}
onIModelAppInit={onIModelAppInit}
enablePerformanceMonitors={true}
/>
```

**Reply by salah-walid** [reply]
The useEffect looks incomplete, or am i missing something ?

**Reply by bsteinbk** [reply]
oops... here is the full useEffect hook
```
  useEffect(()=> {
    const processFrontstageActivation = (args: FrontstageActivatedEventArgs) =>{
      // clear out default tool getter so default tool is not set
      Object.defineProperty(args.activatedFrontstageDef, "_defaultTool", {value: undefined});
      IModelApp?.toolAdmin?.startDefaultTool();
    }

    return FrontstageManager.onFrontstageActivatedEvent.addListener(processFrontstageActivation)
  }, []);
```
